### PR TITLE
Add  TCP and HTTP Drivers

### PR DESCRIPTION
**Pull request by amc-nu**
## Purpose

This PR adds support for TCP and HTTP Drivers using Boost ASIO. Mainly to enable communication with Lidar sensors to obtain diagnostics and perform the setup.

The primary objective of the `tcp_driver` package is to provide wrappers for the
Transmission Control Protocol (TCP) and the Hyper Text Transport Protocol (HTTP)
to allow reliable, ordered delivery of a stream of octets (bytes) between applications
running on hosts (computers) and other devices in a network, such as sensors.

The TCP and HTTP protocols are commonly used by sensors to allow configuration and obtain status reports.

## Assumptions
It is assumed that the source of TCP and HTTP packets might be unreliable. For that reason, the
wrappers implement a timeout option to allow the reconnection and recovery of the connection.

It is also assumed that the output type of the driver is a ROS 2 message.

The HTTP client only implements the GET and POST commands.
PUT and DELETE commands are currently not implemented.

## Implementation

For this, we had to use boost’s `io_context` and not the custom implementation of `transport_drivers::io_context` since that implementation uses boost `io_service`, which was deprecated by [boost.](https://github.com/boostorg/asio/issues/110).

image

### CODE REVIEW

**Comment by wep21** [comment]
@esteve @JWhitleyWork Do you have time to review this PR?

File: .github/workflows/build_test.yml
Comment (esteve) [inline .github/workflows/build_test.yml:14]: Could you reenable rolling? Thanks.

File: .github/workflows/build_test.yml
Comment (esteve) [inline .github/workflows/build_test.yml:32]: Could you open a separate PR for these changes? They don't seem related to supporting TCP and HTTP

File: io_context/include/io_context/common.hpp
Comment (esteve) [inline io_context/include/io_context/common.hpp:20]: What's the rationale for using Boost.Asio and not Asio? Using Boost.Asio brings in a lot of dependencies, whereas Asio only requires a C++11/17 compiler

File: io_context/include/io_context/io_context.hpp
Comment (esteve) [inline io_context/include/io_context/io_context.hpp:31]: If migrating to Boost.Asio from Asio, it'd be best to remove all these comments

File: tcp_driver/design/tcp_driver-design.md
Comment (esteve) [inline tcp_driver/design/tcp_driver-design.md:27]: The correct spelling of the library is Boost.Asio (see https://www.boost.org/doc/libs/1_81_0/doc/html/boost_asio.html)

File: tcp_driver/include/tcp_driver/http_client.hpp
Comment (esteve) [inline tcp_driver/include/tcp_driver/http_client.hpp:22]: How is Beast installed? I saw that it's a header-only library, so it should be added as a vendor package

**Review by esteve — CHANGES_REQUESTED** [review]

File: .github/workflows/build_test.yml
Comment (amc-nu) [reply .github/workflows/build_test.yml]: Addressed [here](https://github.com/ros-drivers/transport_drivers/pull/84/commits/32943067c12da460a98d4792e5cad3914ae8d2b7)

File: .github/workflows/build_test.yml
Comment (amc-nu) [reply .github/workflows/build_test.yml]: Addressed [here](https://github.com/ros-drivers/transport_drivers/pull/84/commits/32943067c12da460a98d4792e5cad3914ae8d2b7)

File: io_context/include/io_context/common.hpp
Comment (amc-nu) [reply io_context/include/io_context/common.hpp]: Addressed [here](https://github.com/ros-drivers/transport_drivers/pull/84/commits/17df5f06ba3cf8f822d9eecc47e820c1dd5dd188)

File: io_context/include/io_context/io_context.hpp
Comment (amc-nu) [reply io_context/include/io_context/io_context.hpp]: Addressed [here](https://github.com/ros-drivers/transport_drivers/pull/84/commits/17df5f06ba3cf8f822d9eecc47e820c1dd5dd188)

File: tcp_driver/design/tcp_driver-design.md
Comment (amc-nu) [reply tcp_driver/design/tcp_driver-design.md]: Addressed [here](https://github.com/ros-drivers/transport_drivers/pull/84/commits/8871afaee4bab3e6d6928efd9bb0038f5365fe3e)

File: tcp_driver/include/tcp_driver/http_client.hpp
Comment (amc-nu) [reply tcp_driver/include/tcp_driver/http_client.hpp:22]: Beast has been included in Boost since version [1.66](https://www.boost.org/users/history/version_1_66_0.html). There is no need to manually install it since [Ubuntu 20.04](https://packages.ubuntu.com/search?suite=focal&arch=any&searchon=all&keywords=boost)

**Comment by amc-nu** [comment]
@esteve Thank you for your kind review. We addressed your comments.
Could you please have a look again?
Thank you

**Comment by amc-nu** [comment]
@esteve We found that starting from Boost >=1.74 and asio >=1.18.0 coredeclaration issues happen nflicts arise. So when trying to use both the `udp_driver` (asio) and `tcp_driver` (Boost.Asio) ition issues happen. I haven't lookit ed thoroughly but appears that either Boost or Asio added a `using` statement in a header causing the problem. In previous versions of Boost and asio this was not a problem (I confirmed no issue on a vanilla docker Ubuntu 20.04 (Boost 1.71 and asio 1.12)).

Switching to Boost.Asio for the `udp_driver` fixes this issue. Is there any reason you want to avoid Boost?

We selected Beast for the HTTP driver since its constantly maintained and widely proven.

**Comment by esteve** [comment]
> @esteve We found that starting from Boost >=1.74 and asio >=1.18.0 coredeclaration issues happen nflicts arise. So when trying to use both the `udp_driver` (asio) and `tcp_driver` (Boost.Asio) ition issues happen. I haven't lookit ed thoroughly but appears that either Boost or Asio added a `using` statement in a header causing the problem. In previous versions of Boost and asio this was not a problem (I confirmed no issue on a vanilla docker Ubuntu 20.04 (Boost 1.71 and asio 1.12)).
> 
> Switching to Boost.Asio for the `udp_driver` fixes this issue. Is there any reason you want to avoid Boost?

Mainly size and complexity, at the time, when `udp_driver` was written the C++11/17 port of Asio was enough, as it's header only. But if Beast requires Boost.Asio I'm in favor of switching to Boost. I'd like to know @JWhitleyWork 's opinion before merging this, though, as he's the current maintainer.

> We selected Beast for the HTTP driver since its constantly maintained and widely proven.

Yeah, looks like it's quite mature.

File: tcp_driver/include/tcp_driver/http_client.hpp
Comment (esteve) [reply tcp_driver/include/tcp_driver/http_client.hpp:22]: Thanks for the explanation

File: .github/workflows/build_test.yml
Comment (esteve) [reply .github/workflows/build_test.yml]: Thanks. Could you move these changes to a separate PR? I think they're useful, just outside of the scope of this PR

**Comment by esteve** [comment]
@amc-nu thanks for addressing my feedback and for the explanation about Boost.Beast. I'm doing another review.

**Comment by amc-nu** [comment]
Hello @esteve Do you have any other comments?
Thank you

**Review by esteve — APPROVED** [review]

**Comment by wep21** [comment]
New tcp drivers are formatted by clang-format while others are formatted by uncrustify. I think it's better to format uniformly. Also, tcp drivers are missing test setting like
```
if(BUILD_TESTING)
  find_package(ament_lint_auto)
  ament_lint_auto_find_test_dependencies()
endif()
```

**Comment by esteve** [comment]
@amc-nu that was it, thanks for addressing my feedback. Given that this is a significant change (replacing Asio with Boost.Asio), I'll wait for a day or two in case @JWhitleyWork has any feedback, and then merge it.

**Comment by amc-nu** [comment]
@wep21 Thank you for your kind comment. I don't have experience formatting with uncrustify. Can you please provide the commands required to do so?

**Comment by esteve** [comment]
@amc-nu if you add the lines that @wep21 pasted, the new code will be tested against the ROS 2 formatting rules and then you can use `ament_uncrustify` to reformat your code if there's any differences.

**Comment by amc-nu** [comment]
@esteve @wep21 Thank you for your comments.
- I enabled Boost as you approved the change. 
- Reformatted the code using uncrustify
- Enable the lining as suggested by @wep21

**Comment by JWhitleyWork** [comment]
Sorry it has taken me so long to get to this. So, here is my feedback:

HTTP is an application-layer protocol. TCP and UDP are transport-layer protocols. HTTP, therefore, does not make sense in this repository. Additionally, the point of using the stand-alone ASIO over the Boost implementation was that the stand-alone ASIO is lightweight and Boost is not. I would be fine accepting the TCP implementation here using the stand-alone ASIO. I am not OK with the HTTP implementation or the use of Boost.

**Comment by esteve** [comment]
@amc-nu @JWhitleyWork I agree about splitting the TCP and HTTP parts for this PR, perhaps the HTTP package could go into another repository.

However, I think it's worth revisiting the decision of `asio` vs `boost.asio`. The `transport_drivers` were spun off from the Velodyne driver developed at Apex.AI. Some of the reasons we chose `asio` originally were because `asio` is already included as a ROS 2 dependency, because of its size and because we didn't need anything more than what `asio` provides, but we didn't need an HTTP transport, for example. If we did at the time, we might have made a different decision.

I just checked the source code for `Boost.Asio`, and it doesn't seem to depend much on the rest of the Boost ecosystem:

```c++
#include <boost/cerrno.hpp>
#include <boost/date_time/posix_time/posix_time_types.hpp>
#include <boost/regex_fwd.hpp>
#include <boost/system/error_code.hpp>
#include <boost/system/system_error.hpp>
#include <boost/version.hpp>
```

So, I'm ok with adding Boost as a dependency. Alternatively, we could make the transports fully header-only and take a templated `io_service` so downstream packages can choose whether to use `asio` or `Boost.Asio`

**Pull request closed without merging**
